### PR TITLE
Unite asset emitting code paths and fix tile parsing

### PR DESCRIPTION
**Pull request by riknoll**
Fixes  microsoft/pxt-arcade#2986
Fixes microsoft/pxt-arcade#3018

This unites all of the asset emitting and all of the parsing except in the monaco fields (which have some special behavior around creating new assets if the name does not exist)

@shakao you should be able to close https://github.com/microsoft/pxt/pull/7855

@jwunderl this might conflict with https://github.com/microsoft/pxt/pull/7858 because getShortIDCore can now return undefined.

I think I tested this pretty thoroughly, but please review carefully because it touches some key stuff

### CODE REVIEW

File: pxtblocks/fields/field_sprite.ts [hunk 1]
```diff
@@ -34,16 +34,9 @@ namespace pxtblockly {
         protected createNewAsset(text?: string): pxt.Asset {
             const project = pxt.react.getTilemapProject();
             if (text) {
-                const match = /^\s*assets\s*\.\s*image\s*`([^`]+)`\s*$/.exec(text);
-                if (match) {
-                    const asset = project.lookupAssetByName(pxt.AssetType.Image, match[1].trim());
-                    if (asset) return asset;
-                    else if (!this.getBlockData()) {
-                        this.isGreyBlock = true;
```
Comment (jwunderl) [inline pxtblocks/fields/field_sprite.ts:42]: Does it not need to handle grey block here anymore?

**Review by jwunderl — COMMENTED** [review]
one comment / question but otherwise LGTM

shouldn't conflict with my pr, already made a change to handle undefined when I saw your comment on the other one: https://github.com/microsoft/pxt/pull/7858/commits/2f77692a2cf5cd80bf125cd510c8d10611a12951

(and in case anyone else always checks longs regexes by hand just to make sure, here's a quick link with a couple test cases I made https://regex101.com/r/XBc5vB/1)

File: pxtblocks/fields/field_sprite.ts (hunk 1 above)
Comment (riknoll) [reply pxtblocks/fields/field_sprite.ts:42]: I think this should get caught by the one below

**Review by jwunderl — APPROVED** [review]

**Comment by riknoll** [comment]
Pushed a quick change to fix python field editors and decompiling from python

File: pxtblocks/fields/field_tileset.ts [hunk 2]
```diff
@@ -94,7 +94,7 @@ namespace pxtblockly {
         getValue() {
             if (this.selectedOption_) {
                 const tile = this.selectedOption_[2];
-                return isGalleryTile(tile) ? tile.id : `assets.tile\`${displayName(tile)}\``
```
Comment (shakao) [inline pxtblocks/fields/field_tileset.ts:97]: can `isGalleryTile` be deleted?

**Review by shakao — APPROVED** [review]

**Pull request merged**